src/anomalib/data/base/datamodule.py

```diff
@@ -110,18 +111,22 @@ def __init__(
 
         self._samples: DataFrame | None = None
 
+        self._is_setup = False  # flag to track if setup has been called from the trainer
+
     def setup(self, stage: str | None = None) -> None:
         """Set up train, validation and test data.
 
         Args:
             stage: str | None:  Train/Val/Test stages.
                 Defaults to ``None``.
         """
-        if not self.is_setup:
+        if not any(hasattr(self, subset) for subset in ["train_data", "val_data", "test_data"]) or not self._is_setup:
```

Comment:
would it be possible to split this condition with an easier-to-read variable names?

Reply:
done